### PR TITLE
[Hardware] [Intel] Enable Multiprocessing and tensor parallel in CPU backend and update documentation 

### PR DESCRIPTION
**Pull request by bigPYJ1151**
This PR enabled vLLM multiprocessing in CPU backend for improving async LLM engine performance and supporting TP.

The main changes include:
- Use utilities from ```vllm.executor.multiproc_worker_utils``` to manage workers in ```CPUExecutor```.
- Enable tensor parallel for the CPU bcakend. 
- Provide a new env ```VLLM_CPU_OMP_THREADS_BIND``` to simplify the OpenMP thread affinity setting, especially for multiple TP workers.
- Transparently separate api_server process and async LLM engine process to avoid the GIL affect.  
- Update related documents.
- Enable a E2E online serving test in CI.

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by bigPYJ1151** [comment]
Hi @WoosukKwon, I think this PR is ready, would you please help to review it if you have time? Thanks!

**Comment by lhtin** [comment]
@bigPYJ1151 Hi, do you have any plan support pipeline parallel in CPU backend?

**Comment by bigPYJ1151** [comment]
@lhtin Not yet. Looks like PP has no significant benefit for CPU device.

**Comment by WoosukKwon** [comment]
@bigPYJ1151 Sorry for the delay. I've been busy with some other tasks. I will review the PR in ~48 hours. Does this sound ok to you?

**Comment by bigPYJ1151** [comment]
@WoosukKwon No problem, just review it when you are available, thanks! :)

File: vllm/distributed/parallel_state.py [hunk 1]
```diff
@@ -289,6 +289,9 @@ def all_reduce(self, input_: torch.Tensor) -> torch.Tensor:
         pynccl_comm = self.pynccl_comm
         if (pynccl_comm is not None and not pynccl_comm.disabled):
             pynccl_comm.all_reduce(input_)
+        elif input_.is_cpu:
+            import intel_extension_for_pytorch as ipex
+            ipex.distributed.all_reduce(input_, group=self.device_group)
```
Comment (WoosukKwon) [inline vllm/distributed/parallel_state.py:294]: Just to double check: The CPU backend uses gloo runtime for distributed inference. Therefore, it basically relies on gloo for collective communications. One exception to this is all-reduce, which we use the custom implementation in IPEX for higher performance. Is my understanding correct?

File: vllm/envs.py [hunk 2]
```diff
@@ -246,6 +247,11 @@ def get_default_config_root():
     "VLLM_CPU_KVCACHE_SPACE":
     lambda: int(os.getenv("VLLM_CPU_KVCACHE_SPACE", "0")),
 
+    # CPU core ids bound by OpenMP threads, e.g., "0-31", "0,1,2",
+    # "0-31,33". CPU cores of different ranks are separated by '|'.
```
Comment (WoosukKwon) [inline vllm/envs.py:251]: Could you clarify that this env is only for the CPU backend?

**Review by WoosukKwon — COMMENTED** [review]
@bigPYJ1151 Sorry for the delay and many thanks for the PR This is amazing!

I really like the overall direction and the efforts you put on this. Just left some questions and minor suggestion for improving the clarity.

File: docs/source/getting_started/cpu-installation.rst [hunk 3]
```diff
@@ -96,11 +102,17 @@ Performance tips
...
+- When using the online serving, it is recommended to reserve 1-2 CPU cores for the serving framework to avoid CPU oversubscription. For example, on a platform with 32 physical CPU cores, reserving CPU 30 and 31 for the framework and using CPU 0-29 for OpenMP:
+
+.. code-block:: console
+
+    $ export VLLM_CPU_KVCACHE_SPACE=40
+    $ export VLLM_CPU_OMP_THREADS_BIND=0-29 
+    $ python3 -m vllm.entrypoints.openai.api_server --model facebook/opt-125m
 
-- If using vLLM CPU backend on a bare-metal machine, it is recommended to disable the hyper-threading.
+- If using vLLM CPU backend on a machine with hyper-threading, it is recommended to bind only one OpenMP thread on each physical CPU core using ``VLLM_CPU_OMP_THREADS_BIND``.
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply docs/source/getting_started/cpu-installation.rst]: Added an example here.

File: docs/source/getting_started/cpu-installation.rst [hunk 4]
```diff
@@ -96,11 +102,17 @@ Performance tips
     $ export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4:$LD_PRELOAD # prepend the library to LD_PRELOAD
     $ python examples/offline_inference.py # run vLLM
 
-- vLLM CPU backend uses OpenMP for thread-parallel computation. If you want the best performance on CPU, it will be very critical to isolate CPU cores for OpenMP threads with other thread pools (like web-service event-loop), to avoid CPU oversubscription. 
+- When using the online serving, it is recommended to reserve 1-2 CPU cores for the serving framework to avoid CPU oversubscription. For example, on a platform with 32 physical CPU cores, reserving CPU 30 and 31 for the framework and using CPU 0-29 for OpenMP:
+
+.. code-block:: console
+
+    $ export VLLM_CPU_KVCACHE_SPACE=40
+    $ export VLLM_CPU_OMP_THREADS_BIND=0-29
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply docs/source/getting_started/cpu-installation.rst:110]: > QQ: Will the API server automatically use CPU 30 and 31?

Yes, CPU 30 and 31 are reserved for non-openMP threads (e.g., python threads, asyncio event loop, ...), and leveraged by OS scheduler automatically.

[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply docs/source/getting_started/cpu-installation.rst:110]: > I really appreciate this simplification. However, can we further set this env variable internally in vLLM so that users don't have to care about it? Just wondering because it's still not super easy to me.
> 
> For example, users may have the following questions:
> 
> 1. Can I use this arg to control the number of CPU cores I'd like to allocate to vLLM?
> 2. How does this arg relate to the vLLM performance? Allocating more CPUs will improve the performance?

Yes, fully automatically setting is the best solution. It requires to detect the topology of CPU cores and memory nodes. We also want to achieve such out-of-box usage.

The ```VLLM_CPU_OMP_THREADS_BIND``` controls the openMP thread behavior of the model inference, including thread number, thread affinity (pin a inference thread on a fixed CPU core), memory allocation policy (only allocate memory from the closest memory node). We have added two performance tips about this arg for platforms with hyper-threading or multi-socket configuration.

For platforms without hyper-threading or multi-socket, allocating more CPUs for model inference will improve the performance theoretically.

File: .buildkite/run-cpu-test.sh [hunk 5]
```diff
@@ -3,8 +3,8 @@
 set -ex
 
 # Try building the docker image
-docker build -t cpu-test -f Dockerfile.cpu .
-docker build --build-arg VLLM_CPU_DISABLE_AVX512="true" -t cpu-test-avx2 -f Dockerfile.cpu .
+numactl -C 48-95 -N 1 docker build -t cpu-test -f Dockerfile.cpu .
+numactl -C 48-95 -N 1 docker build --build-arg VLLM_CPU_DISABLE_AVX512="true" -t cpu-test-avx2 -f Dockerfile.cpu .
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply .buildkite/run-cpu-test.sh:7]: We usually do some test on the same machine, just for limiting the CPU resource usage of the docker build to avoid affect each other.

File: .buildkite/run-cpu-test.sh [hunk 6]
```diff
@@ -13,16 +13,30 @@ remove_docker_container
 
 # Run the image
 docker run -itd --entrypoint /bin/bash -v ~/.cache/huggingface:/root/.cache/huggingface --cpuset-cpus=48-95 \
-  --cpuset-mems=1 --network host -e HF_TOKEN --env VLLM_CPU_KVCACHE_SPACE=4 --name cpu-test cpu-test
+ --cpuset-mems=1 --privileged=true --network host -e HF_TOKEN --env VLLM_CPU_KVCACHE_SPACE=4 --shm-size=4g --name cpu-test cpu-test
 docker run -itd --entrypoint /bin/bash -v ~/.cache/huggingface:/root/.cache/huggingface --cpuset-cpus=48-95 \
-  --cpuset-mems=1 --network host -e HF_TOKEN --env VLLM_CPU_KVCACHE_SPACE=4 --name cpu-test-avx2 cpu-test-avx2
+ --cpuset-mems=1 --privileged=true --network host -e HF_TOKEN --env VLLM_CPU_KVCACHE_SPACE=4 --shm-size=4g --name cpu-test-avx2 cpu-test-avx2
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply .buildkite/run-cpu-test.sh]: Yes, you are right.

File: .buildkite/run-cpu-test.sh [hunk 7]
```diff
@@ -13,16 +13,30 @@ remove_docker_container
...
+  wget -q https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json 
+  timeout 600 bash -c 'until curl localhost:8000/v1/models; do sleep 1; done' || exit 1
+  python3 benchmarks/benchmark_serving.py \
+    --backend vllm \
+    --dataset-name sharegpt \
+    --dataset ./ShareGPT_V3_unfiltered_cleaned_split.json \
+    --model facebook/opt-125m \
+    --num-prompts 20 \
+    --endpoint /v1/completions \
+    --tokenizer facebook/opt-125m"
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply .buildkite/run-cpu-test.sh:42]: From the CI log, it will take less than 20 seconds (~5s for service setup and ~10s for prompts processing), so I think it is acceptable.

File: vllm/worker/cpu_model_runner.py [hunk 8]
```diff
@@ -345,7 +346,7 @@ def prepare_model_input(
             multi_modal_kwargs=multi_modal_kwargs,
         )
 
-    @torch.inference_mode()
+    @torch.no_grad()
```
[the remark this answers is not in the document]
Comment (bigPYJ1151) [reply vllm/worker/cpu_model_runner.py:349]: The ```inference_mode``` does not work with distributed inference on CPU, similar to #1968. I haven't find out root cause, so using ```no_grad``` temporarily.

File: vllm/distributed/parallel_state.py (hunk 1 above)
Comment (bigPYJ1151) [reply vllm/distributed/parallel_state.py:294]: Right

**Comment by bigPYJ1151** [comment]
Hi @WoosukKwon , thanks for your comments! I have updated them, please check, thanks!

File: vllm/worker/cpu_model_runner.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/worker/cpu_model_runner.py:349]: Got it. Thanks for the explanation!

**Comment by WoosukKwon** [comment]
@bigPYJ1151 The PR looks good to me overall. Can you fix the failed CI test? Thanks!

File: .buildkite/run-cpu-test.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply .buildkite/run-cpu-test.sh]: Oh yes I just realized that `num-prompts` is only 20. But I think it takes time to download and process the ShareGPT dataset. What about using `--random` instead? This will generate the random inputs.

**Comment by bigPYJ1151** [comment]
Hi @WoosukKwon, thanks for your feedback! The CPU CI is fixed. The CUDA CI should pass in general but some cases may fail due to random connection errors.

**Review by WoosukKwon — APPROVED** [review]
@bigPYJ1151 LGTM! Thanks a lot for the PR and for addressing my comments!

**Comment by DamonFool** [comment]
Hi @bigPYJ1151 , the CPU target fails to build on my machine.
Could you please take a look at this https://github.com/vllm-project/vllm/pull/6931 ?
Thanks.

File: requirements-cpu.txt [hunk 9]
```diff
@@ -2,6 +2,6 @@
 -r requirements-common.txt
 
 # Dependencies for x86_64 CPUs
-torch == 2.3.1+cpu; platform_machine != "ppc64le"
-torchvision == 0.18.1+cpu; platform_machine != "ppc64le"   # required for the image processor of phi3v, this must be updated alongside torch
+torch == 2.4.0; platform_machine != "ppc64le"
```
Comment (DamonFool) [inline requirements-cpu.txt:5]: Hi @bigPYJ1151 and @WoosukKwon , this change missed the suffix `+cpu` for torch version which leads to the build failure of the CPU target.
Please take a look at https://github.com/vllm-project/vllm/pull/6931 .
Thanks.
